doc/error_handling_guidelines.md

```diff
@@ -0,0 +1,130 @@
+# Azure CLI Error Handling Guidelines
+
+This document aims to provide the guidelines for command group authors to onboard the error handling method and the new error output format.
+
+Previously in Azure CLI, `CLIError` is widely used to wrap the error messages after the different kinds of error types being caught. Now, `CLIError` is deprecated and it is replaced by a set of new designed error types, which will provide more clear structures for error categorizing and also more actionable error outputs for user experience.
+
+For the new commands authoring, here are what need to be done to get on board. __Any PRs not following rules here will be rejected.__ For the existing `CLIError` in command groups, command group owners should schedule to replace them with the new error types.
+1. [__Mandatory__] Use the new designed error types instead of CLIError, detail section [here](#Error-Type)
+2. [__Mandatory__] Follow the error message authoring guidelines, detail section [here](#Error-Message)
+3. [__Recommendation__] Provide recommendations for users to take action, detail section [here](#Error-Recommendation)
+
+
+## Error Type
+
+### Available Error Types
+
+The new designed error types are provided in [azure/cli/core/azclierror.py](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli-core/azure/cli/core/azclierror.py), where the error types are defined in three layers with their structures below. `AzCLIError` is the base layer for all the new defined error types. The second layer includes the main categories (`UserFault`, `ClientError`, `ServiceError`), which can be shown to users and also can be used for Telemetry analysis. __The third layer includes the specific error types, which are the ones command group authors can use when raising errors.__
+
+```
+| -- AzCLIError                                 # [Base Layer]: DO NOT use it in command groups
+|    | -- UserFault                             # [Second Layer]: DO NOT use it in command groups
+|    |    | -- CommandNotFoundError             # [Third Layer]: Can be used in command groups
+|    |    | -- UnrecognizedArgumentError
+|    |    | -- RequiredArgumentMissingError
+|    |    | -- MutuallyExclusiveArgumentError
+|    |    | -- InvalidArgumentValueError
+|    |    | -- ArgumentParseError               # fallback of argument related errors
+|    |    | -- BadRequestError
+|    |    | -- UnauthorizedError
+|    |    | -- ForbiddenError
+|    |    | -- ResourceNotFoundError
+|    |    | -- AzureResponseError               # fallback of response related errors
+|    |    | -- AzureConnectionError
```

Comment:
I am a user. I am angry when you say errors like connection error, response error are my fault. It is Azure's fault.

Reply:
ConnectionError:
Consider the user's network issues, SSL errors etc. Azure should not be blamed.

ResponseError:
which already excludes the 5xx error, it's a fallback of the response related errors

Reply:
When Azure is down, I will also get a connection error?

Reply:
You could catch the case when Azure is down and raise `ServiceInternalError`